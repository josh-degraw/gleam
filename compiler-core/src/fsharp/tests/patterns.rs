use crate::assert_fsharp;

#[test]
fn alternative_patterns() {
    // reassigning name in alternative patterns
    assert_fsharp!(
        r#"
pub fn foo() {
  let duplicate_name = 1

  case 1 {
    1 | 2 -> {
      let duplicate_name = duplicate_name + 1
      duplicate_name
    }
    _ -> 0
  }
}"#
    );
}

#[test]
fn alternative_patterns1() {
    // Alternative patterns with a clause containing vars
    assert_fsharp!(
        r#"
pub fn foo() {
  case Ok(1) {
    Ok(duplicate_name) | Error(duplicate_name) -> duplicate_name
  }
}"#
    );
}

#[test]
fn alternative_patterns2() {
    // Alternative patterns with a guard clause containing vars
    assert_fsharp!(
        r#"
pub fn foo() {
    let duplicate_name = 1

    case 1 {
        1 | 2 if duplicate_name == 1 -> duplicate_name
        _ -> 0
    }
}"#
    );
}

#[test]
fn alternative_patterns3() {
    assert_fsharp!(
        r#"
pub const constant = Ok(1)

pub fn foo(arg) {
  let _ = constant
  case arg {
    _ if arg == constant -> 1
    _ -> 0
  }
}
"#
    );
}

#[test]
fn pattern_as() {
    assert_fsharp!(
        "pub fn a(x) {
  case x {
    Ok(1 as y) -> 1
    _ -> 0
  }
}"
    );
}

#[test]
fn string_prefix_as_pattern_with_multiple_subjects() {
    assert_fsharp!(
        "pub fn a(x) {
  case x, x {
    _, \"a\" as a <> _  -> a
    _, _ -> \"a\"
  }
}"
    );
}

#[test]
fn string_prefix_as_pattern_with_multiple_subjects_and_guard() {
    assert_fsharp!(
        "pub fn a(x) {
  case x, x {
    _, \"a\" as a <> rest if rest == \"a\" -> a
    _, _ -> \"a\"
  }
}"
    );
}

#[test]
fn string_prefix_as_pattern_with_list() {
    assert_fsharp!(
        "pub fn a(x) {
  case x {
    [\"a\" as a <> _, \"b\" as b <> _] -> a <> b
    _ -> \"\"
  }
}"
    );
}

#[test]
fn string_prefix_as_pattern_with_assertion() {
    assert_fsharp!(
        "pub fn a(x) {
  let assert \"a\" as a <> rest = \"wibble\"
  a
}"
    );
}

#[test]
fn assert_pattern_as_last_statement() {
    assert_fsharp!(
        r#"
pub fn go() {
  let assert Error(b) = Error(1)
}
fn int_bind(x) {
    let assert 1 = x
}
pub type Result(a, b) {
  Ok(a)
  Error(b)
}
"#
    );
}

#[test]
fn matching_on_record_with_multiple_constructors() {
    assert_fsharp!(
        "
pub type Step(element, accumulator) {
  Next(element: element, accumulator: accumulator)
  Done
}

fn go() {
  case [] {
      [] ->
        step
        |> equal(Done)

      [h, ..t] -> {
        let assert Next(h2, t2) = step
        h
        |> equal(h2)
        t2
        |> equal(t)
      }
    }
}

fn equal(a, b) {
  a == b
}
"
    );
}
