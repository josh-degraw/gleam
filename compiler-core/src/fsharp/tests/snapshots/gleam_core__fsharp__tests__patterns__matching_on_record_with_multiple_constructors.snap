---
source: compiler-core/src/fsharp/tests/patterns.rs
expression: "\npub type Step(element, accumulator) {\n  Next(element: element, accumulator: accumulator)\n  Done\n}\n\nfn go(step) {\n  case [] {\n      [] ->\n        step\n        |> equal(Done)\n\n      [h, ..t] -> {\n        let assert Next(h2, t2) = step\n        h\n        |> equal(h2)\n        t2\n        |> equal(t)\n      }\n    }\n}\n\nfn equal(a, b) {\n  a == b\n}\n"
---
module rec my.``mod``

type Step<'element, 'accumulator> =
| Done
| Next of element: 'element * accumulator: 'accumulator
    
let private equal<'a> (a: 'a) (b: 'a) = begin
    a = b
end
let private go<'b> (step: Step<'b, list<'b>>) = begin
    match [] with
    | [] ->
        begin
            let _pipe = step
            (equal _pipe Done)
        end
    | h::t ->
        let (Next(h2, t2)) = step
        begin
            let _pipe = h
            (equal _pipe h2)
        end
        begin
            let _pipe = t2
            (equal _pipe t)
        end
end
